src/pudl_archiver/archivers/eia/eiarecs.py

```diff
@@ -0,0 +1,280 @@
+
+import bs4
+
+from pudl_archiver.archivers.classes import (
+    AbstractDatasetArchiver,
+    ArchiveAwaitable,
+    ResourceInfo,
+)
+from pudl_archiver.frictionless import ZipLayout
+from pudl_archiver.utils import retry_async
+
+logger = logging.getLogger(f"catalystcoop.{__name__}")
+
+BASE_URL = "https://www.eia.gov/consumption/residential/data/"
+
+
+@dataclass(frozen=True)
+class TabInfo:
+    """Information needed to archive the links in a tab."""
+
+    url: str
+    name: str
+    year: int
+
+
+class EiaRECSArchiver(AbstractDatasetArchiver):
+    """EIA RECS archiver."""
+
+    name = "eiarecs"
+    base_url = "https://www.eia.gov/consumption/residential/data/2020/"
+
+    async def __get_soup(self, url: str) -> bs4.BeautifulSoup:
+        """Get a BeautifulSoup instance for a URL using our existing session."""
+        response = await retry_async(self.session.get, args=[url])
+        # TODO 2025-02-03: for some reason, lxml fails to grab the closing div
+        # tag for tab content - so we use html.parser, which is slower.
+        return bs4.BeautifulSoup(await response.text(), "html.parser")
+
+    async def get_resources(self) -> ArchiveAwaitable:
+        """Download EIA-RECS resources.
+
+        Looks in the "data" dropdown in the navbar for links to each year.
+        """
+        soup = await self.__get_soup(self.base_url)
+        years = soup.select("div.subnav div.dat_block a")
+        numbered_years = [y for y in years if y.text.strip().lower() != "previous"]
+
+        for year in numbered_years:
+            if self.valid_year(year.text.strip()):
+                yield self.__get_year_resources(
+                    url=urljoin(self.base_url, year["href"]),
+                    year=int(year.text.strip()),
+                )
+
+    async def __get_year_resources(self, url: str, year: int) -> ResourceInfo:
+        """Download all data files for a year.
+
+        Finds links to all available tabs, then dispatches each tab to a
+        handler, which downloads content from the tabs and adds it to the
+        year's zip archive.
+
+        Tab handlers are mostly the same for each tab across the years, but
+        there is an ability to add exceptions when necessary.
+
+        Each year's actual forms are also archived - these are mostly not
+        linked from the tabs themselves, so we go to the main survey page and
+        download them there.
+
+        Args:
+            url: a string that represents the base page for this year
+            year: the actual year number we are archiving
+
+        Returns:
+            ResourceInfo: information about this year's zip file & its contents.
+        """
+        self.logger.info(f"Starting {year}")
+
+        tab_infos = await self.__select_tabs(url)
+
+        # most tabs for most years can be handled the same way
+        tab_handlers = {
+            "housing-characteristics": defaultdict(lambda: self.__get_tab_links),
```

Comment:
In [1997](https://www.eia.gov/consumption/residential/data/1997/)/[1993](https://www.eia.gov/consumption/residential/data/1993/) there's a lot of context on the housing characteristics/consumption & expenditures tabs. We might want to archive the tab contents there too.

At which point we might want to just archive the tab contents for every GD tab - would make this "different defaults for each tab type, plus overrides" logic a lot simpler.

What do you think?